File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第5章方格分割.py
#方格分割  2017年第八届省赛C/C++A组 lanqiaoOJ 644

dir = [(-1, 0), (1, 0), (0, -1), (0, 1)]   #上下左右4个方向
vis = [[1] * 7 for i in range(7)]
cnt = 0

def dfs(x, y):
    global cnt
    if x==0 or y==0 or x==6 or y==6: cnt+=1; return
    # 当前点和对称点都标注访问
    vis[x][y], vis[6-x][6-y] = 0, 0
    for i in range(0, 4):
        newx = x + dir[i][0]        # 新坐标
        newy = y + dir[i][1]
        if newx<0 or newx>6 or newy<0 or newy>6:  continue
        if vis[newx][newy]:  dfs(newx, newy)
    vis[x][y], vis[6 - x][6 - y] = 1, 1 
 
dfs(3, 3)
print(cnt//4)

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第5章方格分割.cpp
//方格分割  2017年第八届省赛C/C++A组 lanqiaoOJ 644

#include<bits/stdc++.h>
using namespace std;
int dx[] = {0, -1, 1,  0, 0};     //上下左右4个方向
int dy[] = {0,  0, 0, -1, 1};
bool vis[10][10];             //标记点是否被访问过
int cnt = 0;
void dfs(int x, int y){
	if(x == 0 || y == 0 || x == 6 || y == 6){cnt++; return;}
	for(int i = 1 ; i <= 4 ; i++){   //上下左右四个方向
		x += dx[i]; y += dy[i];        //走一步
		if(!vis[x][y]){              // 若该点未访问则继续深搜
			vis[x][y] = true;        //  当前的点标注为已访问
			vis[6 - x][6 - y] = true;
			dfs(x, y);         // 继续深搜
			vis[6 - x][6 - y] = false;
			vis[x][y] = false;
		}
		x -= dx[i]; y -= dy[i];
	}
}
int main(){
	vis[3][3] = true;
	dfs(3, 3);
	cout << cnt / 4 << endl;
	return 0;
}
